### PR TITLE
Feature/nvcc nvrtc atomic

### PR DESCRIPTION
**Pull request by maddyscientist**
Cleanup and enablement PR
* Fix nvrtc compilation (must not use `static` shared memory)
* Enable heterogenous atomics for nvc++, clang, and nvrtc (Jitify)
* Remove legacy cub
* Some cleanup and warnings fixes needed for these compilers

### CODE REVIEW

**Comment by weinbe2** [comment]
This PR seems pretty uncontroversial but I don't really have a stake in the changes. My only open question is if this will have any issues with CUDA 11.0, since that's what's officially supported on Summit (and newer flavors of CUDA 11.* have various issues with GDR, etc). Did that get tested?

**Comment by maddyscientist** [comment]
Discussed offline with @weinbe2, but addressing here for posterity:
* No effect with CUDA 11.0: libcu++ was shipped starting with CUDA 10.2, and we've always enabled heterogenous atomics by default on CUDA 11.0+.
* Similarly, no effect on Summit due to the removal of the legacy CUB, since we have utilized the toolkit version of CUB by default for CUDA 11.0+.

File: tests/utils/command_line_params.cpp [hunk 1]
```diff
@@ -145,7 +145,7 @@ QudaTransferType staggered_transfer_type = QUDA_TRANSFER_COARSE_KD;
 // we only actually support 4 here currently
 quda::mgarray<std::array<int, 4>> geo_block_size = {};
 
-#if (CUDA_VERSION >= 10010 && __COMPUTE_CAPABILITY__ >= 700)
+#ifdef QUDA_MMA_AVAILABLE
```
Comment (hummingtree) [inline tests/utils/command_line_params.cpp:148]: Just a thought: we are removing `CUDA_VERSION >= 10010` as we now require CUDA 11.0, should we also remove it in the definition of `QUDA_MMA_AVAILABLE`?

**Review by hummingtree — APPROVED** [review]
This is not controversial and it is a nice clean up.

File: tests/utils/command_line_params.cpp (hunk 1 above)
Comment (weinbe2) [reply tests/utils/command_line_params.cpp:148]: IIRC we still technically support Pascal? I've lost track, tbh.

Comment (mathiaswagner) [reply tests/utils/command_line_params.cpp:148]: Yes, but can and should remove the CUDA_VERSION here.

Comment (maddyscientist) [reply tests/utils/command_line_params.cpp:148]: 7c0a6bd

**Review by weinbe2 — APPROVED** [review]
Verified this compiles with CUDA 11.0 on Summit, we're all good.

File: tests/utils/command_line_params.cpp (hunk 1 above)
Comment (mathiaswagner) [reply tests/utils/command_line_params.cpp:148]: What happens on sm60 here?

**Review by mathiaswagner — APPROVED** [review]

**Pull request merged**